### PR TITLE
Add a `sess` command to MSF and Meterpreter shells

### PR DESCRIPTION
**Pull request by OJ**
## Overview

This PR adds a new command to both the MSF and Metepreter shells is a simpler shortcut that allows for moving around sessions much faster from within the console.
- From inside MSF, `sess <id>` is shorthand for `sessions -i <id>`
- From inside Meterp, `sess <id>` is shorthand for `background; sessions -i <id>`

In the latter case, if the session being switched to is the same id, then no switching happens.
## Sample Run

Typical sessions

```
msf exploit(handler) > sessions

Active sessions
===============

  Id  Type                   Information                           Connection
  --  ----                   -----------                           ----------
  1   meterpreter x86/win32  DESKTOP-5A73R51\oj @ DESKTOP-5A73R51  10.1.10.35:4444 -> 10.1.10.35:64613 (172.16.255.131)
  2   meterpreter x86/win32  WIN-7CH5RT177BA\oj @ WIN-7CH5RT177BA  10.1.10.35:4444 -> 10.1.10.44:49184 (10.1.10.44)

msf exploit(handler) > sessions -i 1
[*] Starting interaction with 1...

meterpreter > machine_id
[+] Machine ID: 3fc9cec59007ffcd84234dfb9aefc1e6
meterpreter > background
[*] Backgrounding session 1...
msf exploit(handler) > sessions -i 2
[*] Starting interaction with 2...

meterpreter > machine_id
[+] Machine ID: b8cbd32fa758033e81eb1143d24948da
meterpreter > background
[*] Backgrounding session 2...
```

There is now a new command in the MSF shell:

```
msf exploit(handler) > sess
Usage: sess <session id>

Interact with the given session ID.
This works the same as: sessions -i <session id>
```

We can now skip the painful `sessions -i` syntax (my personal most common use):

```
msf exploit(handler) > sess 2
[*] Starting interaction with 2...

meterpreter > machine_id
[+] Machine ID: b8cbd32fa758033e81eb1143d24948da
```

This exists in Meterp as well:

```
meterpreter > sess
Usage: sess <session id>

Interact with a different session Id.
This works the same as calling this from the MSF shell: sessions -i <session id>
```

And we can skip from one Meterp session to another!

```
meterpreter > sess 1
[*] Backgrounding session 2...
[*] Starting interaction with 1...

meterpreter > machine_id
[+] Machine ID: 3fc9cec59007ffcd84234dfb9aefc1e6
meterpreter > sess 2
[*] Backgrounding session 1...
[*] Starting interaction with 2...

meterpreter > machine_id
[+] Machine ID: b8cbd32fa758033e81eb1143d24948da
```

And if you try to switch to your current session...

```
meterpreter > sess 2
[*] Session 2 is already interactive.
meterpreter > sess 1000
[*] Backgrounding session 1...
[-] Invalid session identifier: 1000
```

It even works with `-1` as the shortcut:

```
msf exploit(handler) > sess -1
[*] Starting interaction with 2...

meterpreter > sess 1
[*] Backgrounding session 2...
[*] Starting interaction with 1...

meterpreter > sess -1
[*] Backgrounding session 1...
[*] Starting interaction with 2...

meterpreter >
```

My life just got a whole lot better.
## Verification
- [x] Fire up MSF console, and create a handler that can receive Meterpreter sessions.
- [x] Create a matching payload that can hit the handler.
- [x] Launch the payload twice to get two sessions.
- [x] Make sure that `sess` shows help inside the MSF console.
- [x] Make sure that `sess` shows help inside the Meterpreter shell.
- [x] Make sure that `sess <id>` functions the same as `sessions -i <id>` (including with `-1`).
- [x] Make sure that `sess <id>` functions the same as `background; sessions -i <id>` (including with `-1`).
- [x] Interact with session `1`, and in the Meterpreter shell type `sess 1`. No switching should occur.

That should be about it!

### CODE REVIEW

**Comment by OJ** [comment]
I will personally hug anyone who's happy to land this quickly 😍 (I hope that isn't s deterrent).

File: lib/msf/ui/console/command_dispatcher/core.rb [hunk 1]
```diff
@@ -1955,22 +1975,26 @@ def cmd_sessions(*args)
...
+        session = verify_session(sid)
+        if session
+          if session.respond_to?(:response_timeout)
+            last_known_timeout = session.response_timeout
+            session.response_timeout = response_timeout
+          end
+          print_status("Starting interaction with #{session.name}...\n") unless quiet
+          begin
+            self.active_session = session
+            sid = session.interact(driver.input.dup, driver.output)
```
Comment (OJ) [inline lib/msf/ui/console/command_dispatcher/core.rb]: This is the magic line, which stores the next `sid` locally and that is checked in a loop condition.

File: lib/rex/ui/interactive.rb [hunk 2]
```diff
@@ -83,8 +83,13 @@ def interact(user_input, user_output)
       self.completed = true
     end
 
-    # Return whether or not EOF was reached
-    return eof
```
Comment (OJ) [inline lib/rex/ui/interactive.rb]: This bit is the only concern I have. I looked at the code and I couldn't see anywhere that this result was actually relied on, but obviously I could be wrong! So I'd appreciate an expert opinion on this one.

**Comment by bcook-r7** [comment]
Can I bikeshed a little? If you called it 'interact' rather than 'sess', it becomes a little more self-documenting.

And, if 'sessions' and all of its functionality was passed through straight through to the the meterpreter session, there's a certain 'principal of least surprise'.

**Comment by OJ** [comment]
Don't worry, someone was going to mention it :)

Deliberately went with `sess` because everyone knows these things to be sessions already. If people want `sessions` instead, then they're not way off. It makes more sense to me personally to have it called `sess` (it is, after all, the shorter `session`!).

I'll happily go with the majority for the official version, but regardless my local fork will always have `sess` in it :) `interact` doesn't make it much shorter, and retraining the muscle memory will be harder.

Just my $0.02 ;)

**Comment by bcook-r7** [comment]
Good points, the court rests.

**Comment by OJ** [comment]
Oh I'm certainly open to compelling arguments! I promise to keep my tantrums to my private quarters :D

**Comment by bcook-r7** [comment]
Well, 'interact' is the command that at least one other tool uses, so it might help the muscle memory of those folks. Is aliasing both too gross?

**Comment by mubix** [comment]
I have students and I during demos type `session` instead of `sessions` ALL the time, so if that was what was going to happen I'd be jumping up and down. I also like @bcook-r7 's idea for a `interact` command that just shortcuts `sessions -i` but I'm not really seeing much adoptions for `sess` command. Sorry @OJ :/

**Comment by OJ** [comment]
> I'm not really seeing much adoptions for sess

And what sample space are you pulling that from? :)

Like I said, I'll go with the majority, but in my fork there'll always be a `sess` command. `interact` isn't part of any muscle memory, nor does it say "short for session" in my head.

**Comment by OJ** [comment]
BUMP!

I'm happy to have one alias the other if you guys are. Say the magic words and I'll make it so.

**Comment by OJ** [comment]
@bcook-r7 @mubix Sorry to say, but Meterpreter already has an `interact` command which deals with channel interaction. Conflating that with sessions would be a bad idea.

Looks like `sess` stays 😀

**Comment by bcook-r7** [comment]
The only true way to have empirical results its to land it and see who complains, so away it goes.

File: lib/rex/ui/interactive.rb (hunk 2 above)
Comment (bcook-r7) [inline lib/rex/ui/interactive.rb]: I couldn't find a place either. I see _interact checks if the session is stopped directly anyway.

**Comment by OJ** [comment]
Thank you @bcook-r7 ! Appreciate your help dude.

**Comment by bcook-r7** [comment]
## Release Notes

The msfconsole prompt and Meterpreter session prompts now share a common 'sess' command for quickly switching between session IDs directly.

**Comment by kernelsmith** [comment]
you guys do know you could have just aliased these commands right?  I've been aliasing `si` to `sessions -i` for years

**Comment by OJ** [comment]
The original intent was to function differently. Plus the command needed to be implemented in the Meterpreter console as well.

**Comment by kernelsmith** [comment]
well someone just needs to make it available to meterp console too... *not it* ;)

**Comment by kernelsmith** [comment]
I think I actually started doing that at one point, too long ago to remember what happened tho

**Comment by kernelsmith** [comment]
also seems that somewhere along the way, compound command aliases got busted, cuz this used to work in the early days:

**Comment by kernelsmith** [comment]
I have aliasing working in meterpreter, but it's almost identical code to the Msf::Plugin, so I'm looking at a way to assimilate the two to reduce duplicated code

**Pull request merged**